### PR TITLE
Add module support for UpdateTwins2Async

### LINKED ISSUE #554: RegistryManager.UpdateTwins2Async method doesn't work when trying to update a module twin
- **OS, version, SKU and CPU architecture used:** Windows 10 Desktop x64 
- **Application's .NET Target Framework :** net462 
- **Device:** Laptop 
- **SDK version used:** Microsoft.Azure.Devices.1.16.0 

## Description of the issue:
`RegistryManager.UpdateTwins2Async` method doesn't work when trying to update a module twin, instead it updates the device twin itself.
I checked with Fiddler, and the request body indeed lacks the moduleId property. I added it manually and sent the request again and the module twin updated as expected.

## Code sample exhibiting the issue:
```
var twin = new Twin();
twin.ModuleId = "mymod";
twin.DeviceId = "device1";
twin.Tags = new TwinCollection();
twin.Tags["mytag"] = "tag";
var rm = RegistryManager.CreateFromConnectionString("<connstring>");
rm.UpdateTwins2Async(new List<Twin> { twin }, true).Wait();
```

### PR DESCRIPTION
**Pull request by jenoola**
## Checklist
- [X] I have read the [contribution guidelines](https://github.com/Azure/azure-iot-sdk-csharp/blob/master/.github/CONTRIBUTING.md).
- [X] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- [X] This pull-request is submitted against the `master` branch.

## Description of the changes
- Add support for udpating module twins in bulk
- Add end-to-end tests to cover new functionality

## Reference/Link to the issue solved with this PR (if any)
Fixes #554

### CODE REVIEW

File: e2e/test/BulkOperationsE2ETests.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        private readonly string DevicePrefix = $"E2E_{nameof(BulkOperationsE2ETests)}_";
+        private readonly string ModulePrefix = $"E2E_{nameof(BulkOperationsE2ETests)}_";
+
+        private static TestLogging _log;
+        private readonly ConsoleEventListener _listener;
+
+        public BulkOperationsE2ETests()
+        {
+            _listener = TestConfig.StartEventListener();
+            _log = TestLogging.GetInstance();
```
Comment (CIPop) [inline e2e/test/BulkOperationsE2ETests.cs]: Static initializers should never be in a non-static ctor. See other tests for reference.

File: e2e/test/BulkOperationsE2ETests.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        }
+
+        [TestMethod]
+        public async Task BulkOperations_UpdateTwins2_Device()
```
Comment (CIPop) [inline e2e/test/BulkOperationsE2ETests.cs]: Naming: please name tests as "TestComponent_Action_ExpectedResult." See http://osherove.com/blog/2005/4/3/naming-standards-for-unit-tests.html for more details.

File: e2e/test/BulkOperationsE2ETests.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        {
+            await BulkOperations_UpdateTwins2(false).ConfigureAwait(false);
```
Comment (CIPop) [inline e2e/test/BulkOperationsE2ETests.cs]: Always name parameters if it's not clear what they are. `setModule: false` in this case.

File: e2e/test/BulkOperationsE2ETests.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        }
+
+        [TestMethod]
+        public async Task BulkOperations_UpdateTwins2_Module()
+        {
+            await BulkOperations_UpdateTwins2(true).ConfigureAwait(false);
```
Comment (CIPop) [inline e2e/test/BulkOperationsE2ETests.cs]: Same as above.

File: e2e/test/BulkOperationsE2ETests.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        }
+
+        private async Task BulkOperations_UpdateTwins2(bool setModule)
+        {
+            var tagName = Guid.NewGuid().ToString();
+            var tagValue = Guid.NewGuid().ToString();
+
+            TestDevice testDevice = setModule ? null : await TestDevice.GetTestDeviceAsync(DevicePrefix).ConfigureAwait(false);
```
Comment (CIPop) [inline e2e/test/BulkOperationsE2ETests.cs]: This code is very hard to read. Instead of switching between testDevice and testModule, it would have been clearer to write two separate tests.

File: e2e/test/BulkOperationsE2ETests.cs (hunk 1 above)
Comment (CIPop) [reply e2e/test/BulkOperationsE2ETests.cs]: This isn't used - please remove everything that's not used.

File: e2e/test/BulkOperationsE2ETests.cs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+namespace Microsoft.Azure.Devices.E2ETests
+{
+    [TestClass]
+    [TestCategory("IoTHub-E2E")]
+    public class BulkOperationsE2ETests : IDisposable
+    {
+        private readonly string DevicePrefix = $"E2E_{nameof(BulkOperationsE2ETests)}_";
+        private readonly string ModulePrefix = $"E2E_{nameof(BulkOperationsE2ETests)}_";
+
+        public BulkOperationsE2ETests() {}
```
Comment (CIPop) [inline e2e/test/BulkOperationsE2ETests.cs]: Why is this needed?

File: e2e/test/BulkOperationsE2ETests.cs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+                var result = await registryManager.UpdateTwins2Async(new List<Twin> { twin }, true).ConfigureAwait(false);
+                Assert.IsTrue(result.IsSuccessful, $"UpdateTwins2Async error:\n{ResultErrorsToString(result)}");
+
+                Twin twinUpd = await registryManager.GetTwinAsync(testDevice.Id).ConfigureAwait(false);
+
+                Assert.AreEqual(twin.DeviceId, twinUpd.DeviceId, "Device ID changed");
+                Assert.AreEqual(twin.ModuleId, twinUpd.ModuleId, "Module ID changed");
+                Assert.IsNotNull(twinUpd.Tags, "Twin.Tags is null");
+                Assert.IsTrue(twinUpd.Tags.Contains(tagName), "Twin doesn't contain the tag");
+                Assert.AreEqual(twin.Tags[tagName] as string, twinUpd.Tags[tagName] as string, "Tag value changed");
```
Comment (CIPop) [inline e2e/test/BulkOperationsE2ETests.cs]: It's not a good idea to do as string - if it's not string, it'll be `null` which means that AreEqual will pass which is wrong.
Use (string)twin.Tags[tagName] instead to ensure that if the type is not string the code will throw.

File: e2e/test/BulkOperationsE2ETests.cs [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+            TestDevice testDevice = await TestDevice.GetTestDeviceAsync(DevicePrefix).ConfigureAwait(false);
+
+            using (RegistryManager registryManager = RegistryManager.CreateFromConnectionString(Configuration.IoTHub.ConnectionString))
+            {
+                Twin twin = await registryManager.GetTwinAsync(testDevice.Id).ConfigureAwait(false);
+
+                twin.Tags = new TwinCollection();
+                twin.Tags[tagName] = tagValue;
+
+                var result = await registryManager.UpdateTwins2Async(new List<Twin> { twin }, true).ConfigureAwait(false);
```
Comment (CIPop) [inline e2e/test/BulkOperationsE2ETests.cs]: Sorry, I missed this in the initial review: this doesn't seem to test the bug referenced. 
Please add another test which is creating a twin patch for a particular device and one more for a twin patch for a particular device+module (without using `GetTwinAsync` initially).

File: e2e/test/BulkOperationsE2ETests.cs [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        public async Task BulkOperations_UpdateTwins2Device_Ok()
+        {
+            var tagName = Guid.NewGuid().ToString();
+            var tagValue = Guid.NewGuid().ToString();
+
+            TestDevice testDevice = await TestDevice.GetTestDeviceAsync(DevicePrefix).ConfigureAwait(false);
+
+            using (RegistryManager registryManager = RegistryManager.CreateFromConnectionString(Configuration.IoTHub.ConnectionString))
+            {
+                Twin twin = await registryManager.GetTwinAsync(testDevice.Id).ConfigureAwait(false);
```
Comment (CIPop) [inline e2e/test/BulkOperationsE2ETests.cs]: Isn't the code fix below also populating now ModuleId in the twin obtained here? We should then have some asserts that the twin deviceId (and below moduleId) match the ones you expect (TestDevice.Id / TestModule.Id).

Comment (jenoola) [reply e2e/test/BulkOperationsE2ETests.cs]: no, it's gonna be a different device for TestModule, because TestModule generates it's own device

File: iothub/service/src/HttpRegistryManager.cs [hunk 10]
```diff
@@ -828,6 +828,7 @@ static IEnumerable<ExportImportDevice> GenerateExportImportDeviceListForTwinBulk
 
                 var exportImportDevice = new ExportImportDevice();
                 exportImportDevice.Id = twin.DeviceId;
+                exportImportDevice.ModuleId = twin.ModuleId;
```
[the remark this answers is not in the document]
Comment (jenoola) [reply iothub/service/src/HttpRegistryManager.cs]: no, I still need to figure out what to do with other fields. I want to fix customer bug sooner, so this PR is only about supporting module twins

**Comment by CIPop** [comment]
@jenoola - did the tests pass?

**Comment by CIPop** [comment]
https://azure-iot-sdks.visualstudio.com/azure-iot-sdks/_build/results?buildId=6146
